draft/0001-organization.md

```diff
@@ -0,0 +1,272 @@
+
+#### 6.1.2 Successful token response
+
+If the request is valid, and the authorization server successfully issues an access token, the response body SHOULD meet the requirements outlined in [Section 12.2](https://openid.net/specs/openid-connect-core-1_0.html#RefreshTokenResponse) of OpenID Connect.
+
+In addition, the access token MUST be a JWT token containing the following claims:
+
+- `jti`: The unique identifier of the access token.
+- `iss`: The issuer identifier of the authorization server.
+- `sub`: The user identifier.
+- `org`: The unique identifier of the organization.
+- `exp`: The expiration time of the access token.
+- `iat`: The time when the access token was issued.
+- `scope`: The scopes, represented as a whitespace-delimited string, MUST be a subset of the scopes granted to the user.
+
+A non-normative example of the access token:
+
+```json
+{
+  "jti": "abcde",
+  "iss": "https://logto.app",
+  "sub": "12345",
+  "org": "12345",
+  "exp": 1632571200,
+  "iat": 1632484800,
+  "scope": "read:books update:books"
+}
+```
+
+#### 6.1.3. Error response
+
+If the request is invalid, the authorization server SHOULD return an error response in the same format as defined in [Section 5.2](https://www.rfc-editor.org/rfc/rfc6749.html#section-5.2) of OAuth 2.0.
+
+Note that the authorization server SHOULD return the same error response (error code `invalid_request`) for both invalid organization ID and other issues, such as the user not being a member of the organization, to prevent the client from guessing organization information.
+
+### 6.2. Token validation
+
+Before granting access to the specific organization, the resource server MUST validate the access token as follows:
+
+1. The access token MUST be a JWT token.
+2. The access token MUST have a valid signature according to the algorithm defined in the `alg` header parameter and the JSON Web Key Set (JWKS) of the OpenID Provider (OP).
+3. The `iss` claim MUST exactly match the issuer identifier of the OP.
+4. The `org` claim MUST exactly match one of the valid organization identifiers in the resource server.
+5. The `exp` claim MUST be greater than or equal to the current Unix time. The resource server MAY add a reasonable leeway to this time.
+6. The `iat` claim MUST be less than or equal to the current Unix time.
+7. The `scope` claim MUST be a subset of the permissions in the organization template.
+
+## 7. Drawbacks
+
+### 7.1. ID token size
+
+The introduction of new claims may increase the size of the ID token. In a multi-tenant environment, the ID token size may significantly increase if the user has numerous memberships and roles.
+
+To mitigate this issue, the OP MAY choose to include organization claims in the ID token only when the client requests the [Userinfo Endpoint](https://openid.net/specs/openid-connect-core-1_0.html#UserInfo) with the `organization` scope.
+
+### 7.2. Allowed dynamic scopes
```

Comment:
here are some alternatives:

- passing all organization scopes in the initial auth request (i.e. shared with OAuth/OpenID)
- use another parameter for orgnaization scopes

Reply:
this is out-dated. need to specify scopes in the auth request for security